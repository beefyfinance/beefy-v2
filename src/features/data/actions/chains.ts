import { createAsyncThunk } from '@reduxjs/toolkit';
import { ChainConfig } from '../apis/config';
import { getConfigApi } from '../apis/instances';

export interface FulfilledPayload {
  chainConfigs: ChainConfig[];
}

export const fetchChainConfigs = createAsyncThunk<FulfilledPayload>(
  'chains/fetchChainConfigs',
  async () => {
    const api = getConfigApi();
    const chainConfigs = await api.fetchChainConfigs();
    return { chainConfigs };
  }
);
