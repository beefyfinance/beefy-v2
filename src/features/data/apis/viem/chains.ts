import type { Chain } from 'viem/chains';
import type { ChainEntity } from '../../entities/chain';

export function buildViemChain(chain: ChainEntity): Chain {
  return {
    id: chain.networkChainId,
    name: chain.name,
    nativeCurrency: {
      decimals: 18,
      name: chain.walletSettings.nativeCurrency.name,
      symbol: chain.walletSettings.nativeCurrency.symbol,
    },
    rpcUrls: {
      public: { http: chain.rpc },
      default: { http: chain.rpc },
    },
    blockExplorers: {
      default: { name: `${chain.name} Explorer`, url: chain.explorerUrl },
    },
    contracts: {
      multicall3: {
        address: chain.multicall3Address,
      },
    },
  };
}
