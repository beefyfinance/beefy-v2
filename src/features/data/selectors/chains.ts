import { createSelector } from '@reduxjs/toolkit';
import { BeefyState } from '../../../redux-types';
import { ChainEntity } from '../entities/chain';

export const selectChainById = createSelector(
  // get a tiny bit of the data
  (state: BeefyState) => state.entities.chains.byId,
  // get the user passed ID
  (_: BeefyState, chainId: ChainEntity['id']) => chainId,
  // last function receives previous function outputs as parameters
  (chainsById, chainId) => {
    if (chainsById[chainId] === undefined) {
      throw new Error(`selectChainById: Unknown chain id ${chainId}`);
    }
    return chainsById[chainId];
  }
);

export const selectAllChains = createSelector(
  // get a tiny bit of the data
  (state: BeefyState) => state.entities.chains.allIds,
  // get a tiny bit of the data
  (state: BeefyState) => state.entities.chains.byId,
  // last function receives previous function outputs as parameters
  (allIds, byId) => allIds.map(id => byId[id])
);

export const selectAllChainIds = (state: BeefyState) => state.entities.chains.allIds;
