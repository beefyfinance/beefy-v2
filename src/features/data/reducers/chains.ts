import { createSlice } from '@reduxjs/toolkit';
import { fetchChainConfigs } from '../actions/chains';
import { ChainEntity } from '../entities/chain';
import { NormalizedEntity } from '../utils/normalized-entity';

/**
 * State containing Vault infos
 */
export type ChainsState = NormalizedEntity<ChainEntity>;
export const initialChainsState: ChainsState = {
  byId: {},
  allIds: [],
};

export const chainsSlice = createSlice({
  name: 'chains',
  initialState: initialChainsState,
  reducers: {
    // standard reducer logic, with auto-generated action types per reducer
  },
  extraReducers: builder => {
    builder.addCase(fetchChainConfigs.fulfilled, (sliceState, action) => {
      for (const chainConfig of action.payload.chainConfigs) {
        // we already know this chain
        if (chainConfig.id in sliceState.byId) {
          continue;
        }
        // for now, both ChainEntity and ChainConfig have somewhat compatible types
        // we just rename the chainId to networkChainId to avoid confusion
        const { chainId: _, ...chainConfigData } = chainConfig;
        const chain: ChainEntity = {
          ...chainConfigData,
          networkChainId: chainConfig.chainId,
        };

        sliceState.byId[chain.id] = chain;
        sliceState.allIds.push(chain.id);
      }
    });
  },
});
