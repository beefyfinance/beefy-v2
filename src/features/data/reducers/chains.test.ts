import {
  fetchChainConfigs,
  FulfilledPayload as FetchConfigsActionPayload,
} from '../actions/chains';
import { chainsSlice, initialChainsState } from './chains';

describe('Chains slice tests', () => {
  it('should update state on fulfilled chain config list', () => {
    const payload: FetchConfigsActionPayload = {
      chainConfigs: [
        {
          id: 'arbitrum',
          name: 'Arbitrum',
          chainId: 42161,
          rpc: ['https://arb1.arbitrum.io/rpc'],
          explorerUrl: 'https://arbiscan.io',
          multicallAddress: '0x13aD51a6664973EbD0749a7c84939d973F247921',
          supportedWallets: ['injected', 'custom-coinbase', 'custom-wallet-connect'],
          providerName: 'Arbitrum',
          walletSettings: {
            chainId: `0x${parseInt('42161', 10).toString(16)}`,
            chainName: 'Arbitrum One',
            nativeCurrency: {
              name: 'ETH',
              symbol: 'ETH',
              decimals: 18,
            },
            rpcUrls: ['https://arb1.arbitrum.io/rpc'],
            blockExplorerUrls: ['https://arbiscan.io/'],
          },
          stableCoins: ['USDC', 'USDT', 'MIM'],
        },
        {
          id: 'celo',
          name: 'Celo',
          chainId: 42220,
          rpc: ['https://forno.celo.org'],
          explorerUrl: 'https://explorer.celo.org',
          multicallAddress: '0xa9E6E271b27b20F65394914f8784B3B860dBd259',
          supportedWallets: ['injected', 'custom-wallet-connect'],
          providerName: 'Celo',
          walletSettings: {
            chainId: `0x${parseInt('42220', 10).toString(16)}`,
            chainName: 'Celo',
            nativeCurrency: {
              name: 'CELO',
              symbol: 'CELO',
              decimals: 18,
            },
            rpcUrls: ['https://forno.celo.org'],
            blockExplorerUrls: ['https://explorer.celo.org/'],
          },
          stableCoins: ['cUSD', 'cEUR', 'DAI'],
        },
      ],
    };
    const action = { type: fetchChainConfigs.fulfilled, payload: payload };
    const state = chainsSlice.reducer(initialChainsState, action);
    expect(state).toMatchSnapshot();

    // getting the same vaults don't update the state object
    const beforeReDispatch = Object.values(state.byId)[0];
    const newState = chainsSlice.reducer(state, action);
    const afterReDispatch = newState.byId[beforeReDispatch.id];
    expect(beforeReDispatch).toBe(afterReDispatch);
  });
});
