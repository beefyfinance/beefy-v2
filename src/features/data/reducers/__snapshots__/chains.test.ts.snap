// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chains slice tests should update state on fulfilled chain config list 1`] = `
Object {
  "allIds": Array [
    "arbitrum",
    "celo",
  ],
  "byId": Object {
    "arbitrum": Object {
      "explorerUrl": "https://arbiscan.io",
      "id": "arbitrum",
      "multicallAddress": "0x13aD51a6664973EbD0749a7c84939d973F247921",
      "name": "Arbitrum",
      "networkChainId": 42161,
      "providerName": "Arbitrum",
      "rpc": Array [
        "https://arb1.arbitrum.io/rpc",
      ],
      "stableCoins": Array [
        "USDC",
        "USDT",
        "MIM",
      ],
      "supportedWallets": Array [
        "injected",
        "custom-coinbase",
        "custom-wallet-connect",
      ],
      "walletSettings": Object {
        "blockExplorerUrls": Array [
          "https://arbiscan.io/",
        ],
        "chainId": "0xa4b1",
        "chainName": "Arbitrum One",
        "nativeCurrency": Object {
          "decimals": 18,
          "name": "ETH",
          "symbol": "ETH",
        },
        "rpcUrls": Array [
          "https://arb1.arbitrum.io/rpc",
        ],
      },
    },
    "celo": Object {
      "explorerUrl": "https://explorer.celo.org",
      "id": "celo",
      "multicallAddress": "0xa9E6E271b27b20F65394914f8784B3B860dBd259",
      "name": "Celo",
      "networkChainId": 42220,
      "providerName": "Celo",
      "rpc": Array [
        "https://forno.celo.org",
      ],
      "stableCoins": Array [
        "cUSD",
        "cEUR",
        "DAI",
      ],
      "supportedWallets": Array [
        "injected",
        "custom-wallet-connect",
      ],
      "walletSettings": Object {
        "blockExplorerUrls": Array [
          "https://explorer.celo.org/",
        ],
        "chainId": "0xa4ec",
        "chainName": "Celo",
        "nativeCurrency": Object {
          "decimals": 18,
          "name": "CELO",
          "symbol": "CELO",
        },
        "rpcUrls": Array [
          "https://forno.celo.org",
        ],
      },
    },
  },
}
`;
